crypto/serializer/deserializer_lib.c

```diff
@@ -0,0 +1,333 @@
+
+    ok = deser_process(NULL, &data);
+
+    /* Clear any cached passphrase */
+    OPENSSL_free(ctx->cached_passphrase);
+    ctx->cached_passphrase = NULL;
+    ctx->cached_passphrase_len = 0;
+    return ok;
+}
+
+#ifndef OPENSSL_NO_STDIO
+static BIO *bio_from_file(FILE *fp)
+{
+    BIO *b;
+
+    if ((b = BIO_new(BIO_s_file())) == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_BIO_LIB);
+        return NULL;
+    }
+    BIO_set_fp(b, fp, BIO_NOCLOSE);
+    return b;
+}
+
+int OSSL_DESERIALIZER_from_fp(OSSL_DESERIALIZER_CTX *ctx, FILE *fp)
+{
+    BIO *b = bio_from_file(fp);
+    int ret = 0;
+
+    if (b != NULL)
+        ret = OSSL_DESERIALIZER_from_bio(ctx, b);
+
+    BIO_free(b);
+    return ret;
+}
+#endif
+
+int OSSL_DESERIALIZER_CTX_add_deserializer(OSSL_DESERIALIZER_CTX *ctx,
+                                           OSSL_DESERIALIZER *deser)
+{
+    OSSL_DESERIALIZER_INSTANCE *deser_inst = NULL;
+    const OSSL_PROVIDER *prov = NULL;
+    OSSL_PARAM params[2];
+    void *provctx = NULL;
+
+    if (!ossl_assert(ctx != NULL) || !ossl_assert(deser != NULL)) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_PASSED_NULL_PARAMETER);
+        return 0;
+    }
+
+    if (deser->get_params == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER,
+                  OSSL_DESERIALIZER_R_MISSING_GET_PARAMS);
+        return 0;
+    }
+
+    if (ctx->deser_insts == NULL
+        && (ctx->deser_insts =
+            sk_OSSL_DESERIALIZER_INSTANCE_new_null()) == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    if ((deser_inst = OPENSSL_zalloc(sizeof(*deser_inst))) == NULL) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_MALLOC_FAILURE);
+        return 0;
+    }
+    if (!OSSL_DESERIALIZER_up_ref(deser)) {
+        ERR_raise(ERR_LIB_OSSL_DESERIALIZER, ERR_R_INTERNAL_ERROR);
+        goto err;
+    }
+    deser_inst->deser = deser;
+
+    prov = OSSL_DESERIALIZER_provider(deser_inst->deser);
+    provctx = OSSL_PROVIDER_get0_provider_ctx(prov);
+
+    /* Cache the input type for this serializer */
+    params[0] =
+        OSSL_PARAM_construct_utf8_ptr(OSSL_DESERIALIZER_PARAM_INPUT_TYPE,
+                                      (char **)&deser_inst->input_type, 0);
+    params[1] = OSSL_PARAM_construct_end();
+
+    if (!deser_inst->deser->get_params(params)
+        || !OSSL_PARAM_modified(&params[0]))
+        goto err;
+
+    if ((deser_inst->deserctx = deser_inst->deser->newctx(provctx))
+        == NULL)
+        goto err;
+
+    if (sk_OSSL_DESERIALIZER_INSTANCE_push(ctx->deser_insts, deser_inst) <= 0)
+        goto err;
+
+    return 1;
+ err:
+    if (deser_inst != NULL) {
+        if (deser_inst->deser != NULL)
+            deser_inst->deser->freectx(deser_inst->deserctx);
+        OSSL_DESERIALIZER_free(deser_inst->deser);
+        OPENSSL_free(deser_inst);
+    }
+    return 0;
+}
+
+int OSSL_DESERIALIZER_CTX_add_extra(OSSL_DESERIALIZER_CTX *ctx,
```

Comment:
I would just like to say that this function is not easy to understand. I will come back to it when I have browsed further, and have figured out where it is used.

Reply:
Yeah...  although, those names are already a mouthful, I tried to keep it short.  Ideas are certainly welcome.

Reply:
Perhaps use "DESER" like you do for the variables?  (And "_instance" as a prefix can probably just be removed.)

Reply:
So prefix everything with `OSSL_DESER` instead of `OSSL_DESERIALIZE`...  for consistency, we should probably change `OSSL_SERIALIZE` to `OSSL_SER`, then...  it doesn't strike me as the most intuitive prefixes, but it is a viable option if we can all agree that the less readable trumps the mouthful.

Reply:
(I have some refactoring thoughts for the serializer API as well, so a name change would simply become part of that)

Reply:
Just offering a suggestion, whatever you guys decide.

Reply:
FWIW, in other software (http://pywebsvcs.sourceforge.net/zsi.html) I used parse as the opposite of serialize.

Reply:
Frankly, I cannot pronounce serializer deserializer without breaking my tongue :smile:
Have you considered exporter/importer?

Reply:
@richsalz, I didn't mean to disqualify your suggestion. I find the function names I came up with excruciatingly long to type, but the naming was stuck in my mind, so I do appreciate suggestions to get unstuck.

Serialize / parse sounds like an odd pair, though. But, makes me think of pack / unpack as an option (although I would probably confuse the two, so maybe not the best choice).

@t8m, export / import could actually be viable! OSSL_EXPORT and OSSL_IMPORT, I kinda like the ring of that. However, it might be confusing when we thing of the export / import functionality in KEYMGMT... Still, that might be a possibility. 
(also, since the KEYMGMT export / import functionality convert internal data to / from OSSL_PARAM, it is essentially a serialization / deserialization... That might help)

Reply:
No need to apologize, glad to help unstick.  Just make sure it also didn't dislodge any perl knowledge :)

Reply:
Hah! No chance

Reply:
I'd prefer the shorter name too.  I don't think most users would have a problem with OSSL_IMPORT/EXPORT.  The KEYMGMT import/export is mostly hidden in the provider boundary.

Kind of similar to the user facing EVP_*_get_params calls versus the get_params calls providers use.